src/rgw/rgw_rados.h

```diff
@@ -2873,6 +2873,9 @@ class RGWRados : public AdminSocketHook
       int write_meta(uint64_t size, uint64_t accounted_size,
                      map<std::string, bufferlist>& attrs);
       int write_data(const char *data, uint64_t ofs, uint64_t len, bool exclusive);
+      const struct req_state* get_req_state() {
+        return (const struct req_state*)(target->get_ctx().user_ctx);
+      }
```

Comment:
i think this piece is going to need more work

first, i didn't see anywhere that actually initialized RGWObjectCtx::user_ctx. but more importantly, there are a lot of places where we do bucket index operations that aren't part of a request (stuff like radosgw-admin commands, object lifecycle expiration, multisite sync, etc.), so we're not guaranteed to have a req_state

are we interested in tracing these background operations too? if so, the code paths that don't have an associated req_state would need to fake up some kind of unique req_id string to pass into tracepoint() instead

either way, we'll need to pass the req_state's req_id string into _do_write_meta() somehow - preferably as a `const char*` instead of `req_state*`, and as an extra argument to either _do_write_meta() or Write's constructor instead of coupling it with RGWObjectCtx

Reply:
@cbodley 
Yes, RGWObjectCtxx::user_ctx may be NULL. All write_meta() calls are in rgw_op.cc and rgw_rados.cc. Every place where write_meta() is called, there must be a related req_state.

req_state is initialized in process_request().
```c++
  struct req_state rstate(g_ceph_context, &rgw_env, &userinfo);    
  struct req_state *s = &rstate;    
  RGWObjectCtx rados_ctx(store, s);  // s => user_ctx   
  ...     
  s->req_id = store->unique_id(req->id);
```
Now only put objects related bi updates are concerned about.

As RGWObjectCtx already has req_state/req_id info, do we really need to add a req_id item in Write?

Reply:
ok, thanks for pointing out where process_request() initializes the user_ctx. that rados_ctx gets stored in `s->obj_ctx` and used throughout rgw_op.cc

but there are a lot of other cases where RGWObjectCtx is initialized on the stack without that user_ctx argument, so we'll still have to make the tracepoints in `_do_write_meta()` deal with the cases where this `get_req_state()` function returns null

Reply:
thinking more about `void* RGWObjectCtx::user_ctx`.. now that we have code that requires it to be a `req_state*` if non-null, then its name and type really should be changed to reflect that. can you please change `void *user_ctx` to `req_state *s`, and also make the second `RGWObjectCtx` constructor `explicit`?

Reply:
@cbodley Done